test/Microsoft.DotNet.Tools.Tests.Utilities/TempFileSystem/DisposableFile.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+namespace Microsoft.DotNet.Tools.Test.Utilities
+{
+    public sealed class DisposableFile : TempFile, IDisposable
+    {
+        public DisposableFile(string path)
+            : base(path)
+        {
+        }
+
+        public DisposableFile(string prefix = null, string extension = null, string directory = null, string callerSourcePath = null, int callerLineNumber = 0)
+            : base(prefix, extension, directory, callerSourcePath, callerLineNumber)
+        {
+        }
+
+        public void Dispose()
+        {
+            if (Path != null && File.Exists(Path))
+            {
+                try
+                {
+                    File.Delete(Path);
+                }
+                catch (UnauthorizedAccessException)
+                {
+                    try
+                    {
+                        // the file might still be memory-mapped, delete on close:
+                        DeleteFileOnClose(Path);
+                    }
+                    catch (IOException ex)
+                    {
+                        throw new InvalidOperationException(string.Format(@"
+The file '{0}' seems to have been opened in a way that prevents us from deleting it on close.
+Is the file loaded as an assembly (e.g. via Assembly.LoadFile)?
+ 
+{1}: {2}", Path, ex.GetType().Name, ex.Message), ex);
+                    }
+                    catch (UnauthorizedAccessException)
+                    {
+                        //  We should ignore this exception if we got it the second time, 
+                        //  the most important reason is that the file has already been 
+                        //  scheduled for deletion and will be deleted when all handles 
+                        //  are closed.
+                    }
+                }
+            }
+        }
+
+        [DllImport("kernel32.dll", PreserveSig = false)]
```

Comment:
`TempDirectory` functionality has some dependency on `DisposableFile`, so I just brought this code over from Roslyn and did not try to fix this. Since this is test infrastructure I think we have low risk not fixing this now.